File: crates/adi-ui/src/button.rs
//! The button — and the reference for how a component in this crate is shaped: an enum per
//! axis of variation, each arm owning one complete class literal, and the assembled string
//! handed to a single element.

use leptos::prelude::*;

use crate::icon::{Icon, IconSize, Lucide};
use crate::merge;

/// What the button is *for*, which is the only thing that decides how it looks.
///
/// `design/DESIGN.md` §2.4: one filled orange per screen. [`Primary`](Self::Primary) is that
/// orange and a screen gets one; when the orange is already spent — an update button in the
/// bar, a running dot the page is about — the page's main action is [`Strong`](Self::Strong),
/// an ink fill. Everything else recedes into the translucent default.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub enum ButtonVariant {
    /// Most buttons: a translucent fill on whatever surface it sits on.
    #[default]
    Default,
    /// The one action a screen exists for — Send, Save, Update. Orange, and one per screen.
    Primary,
    /// The page's main action when orange is taken: ink on the page.
    Strong,
    /// Present but quiet — no fill until you point at it. Cancel, tertiary controls.
    Ghost,
    /// Destructive. Red text on the default fill; the word says the rest.
    Danger,
    /// Reads as a link, behaves as a button. For an action inside prose or a meta line. Never
    /// orange: links inside the app take the ink around them (§3).
    Link,
}

impl ButtonVariant {
    /// The complete class list for this variant — written out per arm rather than composed,
    /// because Tailwind only finds classes it can read as literals in the source.
    #[must_use]
    pub fn classes(self) -> &'static str {
        match self {
            Self::Default => "bg-btn text-ink hover:bg-btn-hover",
            Self::Primary => "bg-accent text-on-accent hover:bg-accent-hover",
            Self::Strong => "bg-ink text-bg hover:bg-white",
            Self::Ghost => "bg-transparent text-ink-2 hover:bg-hover hover:text-ink",
            Self::Danger => "bg-btn text-err hover:bg-btn-hover",
            Self::Link => {
                "bg-transparent px-0 text-ink-2 underline decoration-ink-3 \
                 underline-offset-[3px] hover:text-ink"
            }
        }
    }
}

/// Button height. The default is the spec's `7px 14px`; `Small` is for a control that sits
/// inside a table row or a 48px bar.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub enum ButtonSize {
    /// Row scale: `5px 10px`, 13px.
    Small,
    /// The default: `7px 14px`, 13.5px.
    #[default]
    Medium,
}

impl ButtonSize {
    /// Padding and type size for this step.
    #[must_use]
    pub fn classes(self) -> &'static str {
        match self {
            Self::Small => "gap-1 px-2.5 py-[5px] text-small",
            Self::Medium => "gap-1.5 px-3.5 py-[7px] text-row",
        }
    }

    /// The glyph at this step: 14 in a row-scale button, 16 otherwise (DESIGN.md §9).
    #[must_use]
    pub fn icon_size(self) -> IconSize {
        match self {
            Self::Small => IconSize::Sm,
            Self::Medium => IconSize::Md,
        }
    }
}

/// Shared by every variant: the box, the type, and the states. Focus is the same quiet ring
/// the design system puts on every `:focus-visible` — never orange (§3).
const BASE: &str = "inline-flex items-center justify-center whitespace-nowrap rounded-md \
                    font-medium leading-[1.2] transition-[background-color,color] duration-100 \
                    cursor-pointer select-none \
                    focus-visible:outline-[1.5px] focus-visible:outline-offset-2 \
                    focus-visible:outline-focus \
                    disabled:cursor-not-allowed disabled:opacity-50";

/// A button.
///
/// Event handlers attach the ordinary Leptos way — `<Button on:click=…>` lands on the
/// underlying `<button>`, so this component has no callback prop of its own.
///
/// ```ignore
/// <Button variant=ButtonVariant::Danger size=ButtonSize::Small on:click=drop_it>
///     "Release"
/// </Button>
/// ```
#[component]
pub fn Button(
    #[prop(optional)] variant: ButtonVariant,
    #[prop(optional)] size: ButtonSize,
    /// Rendered as `type="submit"` — the one form-related behaviour worth a prop, since a
    /// bare `<button>` inside a `<form>` submits it whether you meant it to or not.
    #[prop(optional)]
    submit: bool,
    #[prop(optional, into)] disabled: Signal<bool>,
    /// A glyph before the label. Drawn in `currentColor` at the size the button picks, which
    /// is what keeps an icon button on-theme through every variant without the call site
    /// knowing anything. A button with an icon and no children is a square icon button; give
    /// it an `aria-label` when it has no words.
    #[prop(optional)]
    icon: Option<Lucide>,
    /// Extra utilities from the call site: layout, width, margin.
    #[prop(optional, into)]
    class: String,
    #[prop(optional)] children: Option<Children>,
) -> impl IntoView {
    let own = format!("{BASE} {} {}", variant.classes(), size.classes());
    view! {
        <button
            class=merge(&own, class)
            type=if submit { "submit" } else { "button" }
            disabled=move || disabled.get()
        >
            {icon.map(|icon| view! { <Icon icon=icon size=size.icon_size()/> })}
            {children.map(|c| c())}
        </button>
    }
}

File: design/elements/button.js
// `<adi-button>` — DESIGN.md §6 "Button", and §2.4's one orange.
//
//   <adi-button>Cancel</adi-button>
//   <adi-button variant="primary" icon="arrow-up">Send</adi-button>   <!-- the screen's one orange -->
//   <adi-button variant="strong">Save</adi-button>                    <!-- ink fill, when the orange is spent -->
//   <adi-button variant="quiet" href="/extended/settings">Settings</adi-button>
//   <adi-button square icon="ellipsis" label="Row actions"></adi-button>
//
// A naming clash worth knowing about: adi-css spells the ink fill `.adi-btn--primary` and the
// orange `.adi-btn--accent`. DESIGN.md §6 names them the other way round, and this library
// follows the rulebook — `primary` here is the orange, `strong` is the ink fill.

import { AdiElement, define, sheet } from "./base.js";
import "./icon.js";

/** Every fill a button has (§6). `primary` is the screen's one orange; nothing else is. */
const VARIANTS = ["default", "primary", "strong", "quiet", "link", "danger"];

class AdiButton extends AdiElement {
  static observedAttributes = ["variant", "size", "icon", "icon-end", "label", "disabled", "href"];

  static sheet = sheet(`
    :host { display: inline-flex; vertical-align: middle; }
    :host([full]) { display: flex; }
    :host([full]) .btn { width: 100%; }

    .btn {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      gap: 6px;
      padding: 7px 14px;
      border-radius: var(--r);
      background: var(--btn);
      color: var(--ink);
      font-size: var(--fs-ui-sm);
      font-weight: 500;
      line-height: 1.2;
      white-space: nowrap;
      cursor: pointer;
      transition: background var(--transition), color var(--transition);
    }
    .btn:hover { background: var(--btn-hover); }

    :host([variant="primary"]) .btn { background: var(--accent); color: var(--on-accent); }
    :host([variant="primary"]) .btn:hover { background: var(--accent-hover); }

    :host([variant="strong"]) .btn { background: var(--ink); color: var(--bg); }
    /* Lifted rather than swapped for a lighter ink: there is no token for "one step above
       --ink", and §8 forbids inventing one here (adi-css hardcodes #fff at this spot). */
    :host([variant="strong"]) .btn:hover { filter: brightness(1.08); }

    :host([variant="quiet"]) .btn { background: none; color: var(--ink-2); }
    :host([variant="quiet"]) .btn:hover { background: var(--bg-hover); color: var(--ink); }

    /* An action written as a word in a line of text — in a panel head, after a value. */
    :host([variant="link"]) .btn {
      padding: 3px 6px;
      background: none;
      color: var(--ink-2);
      font-weight: 400;
      font-size: var(--fs-small);
    }
    :host([variant="link"]) .btn:hover { background: var(--bg-hover); color: var(--ink); }

    /* Destructive: red text, whatever fill it has (§3). */
    :host([variant="danger"]) .btn { color: var(--err); }

    :host([size="sm"]) .btn { padding: 5px 10px; font-size: var(--fs-small); }

    /* The square that holds one icon, for a bar, a panel head or a row. */
    :host([square]) .btn {
      width: 28px;
      height: 28px;
      padding: 0;
      background: none;
      color: var(--ink-3);
    }
    :host([square][size="sm"]) .btn { width: 24px; height: 24px; }
    :host([square]) .btn:hover { background: var(--bg-hover); color: var(--ink); }

    :host([disabled]) .btn { opacity: .5; cursor: default; pointer-events: none; }
  `);

  template() {
    // `<a>` or `<button>` is settled once, on the first build: a control that changes from one
    // to the other under an attribute is not a thing any caller wants, and the branch would
    // have to survive every update below.
    const tag = this.hasAttribute("href") ? "a" : "button";
    const type = tag === "button" ? ' type="button"' : "";
    return `
      <${tag} class="btn" part="button"${type}>
        <adi-icon part="icon" hidden></adi-icon>
        <slot></slot>
        <adi-icon part="icon-end" hidden></adi-icon>
      </${tag}>
    `;
  }

  update() {
    const control = this.$(".btn");
    const size = this.getAttribute("size") === "sm" ? 14 : 16;
    const lead = this.$('[part="icon"]');
    const trail = this.$('[part="icon-end"]');

    for (const [node, name] of [
      [lead, this.getAttribute("icon")],
      [trail, this.getAttribute("icon-end")],
    ]) {
      node.hidden = !name;
      if (name) {
        node.setAttribute("name", name);
        node.setAttribute("size", size);
      }
    }

    // A square button has no visible words in it, so its name has to come from somewhere (§9).
    const label = this.getAttribute("label");
    if (label) control.setAttribute("aria-label", label);
    else control.removeAttribute("aria-label");
    if (label && !this.hasAttribute("title")) control.title = label;

    const disabled = this.hasAttribute("disabled");
    if (control.tagName === "BUTTON") {
      control.disabled = disabled;
    } else {
      control.href = this.attr("href");
      control.setAttribute("aria-disabled", String(disabled));
      if (this.hasAttribute("target")) control.target = this.attr("target");
    }
  }
}

define("adi-button", AdiButton);

export { AdiButton, VARIANTS };
